x/src/lint/engine/runner.rs

```diff
@@ -0,0 +1,183 @@
+// Copyright (c) The Libra Core Contributors
+// SPDX-License-Identifier: Apache-2.0
+
+use crate::lint::engine::prelude::*;
+use once_cell::sync::OnceCell;
+use std::{
+    ffi::OsStr,
+    path::Path,
+    process::{Command, Stdio},
+};
+
+/// Configuration for the lint engine.
+#[derive(Clone, Debug)]
+pub struct LintEngineConfig<'cfg> {
+    project_root: &'cfg Path,
+    content_linters: Option<&'cfg [&'cfg dyn ContentLinter]>,
+    fail_fast: bool,
+}
```

Comment:
Just a note that structuring the lifetime like this limits the linters that can be added, as if the project_root is a path that has `'static` then the list of linters you provide must also be `'static`. Is that something you're looking to enforce or would it make sense to have another lifetime param for that?

Reply:
`'cfg` is a [covariant lifetime parameter](https://doc.rust-lang.org/nomicon/subtyping.html) so it should automatically be shrunk down to the smallest lifetime involved. Multiple params are only necessary with invariant lifetime params, which show up with `&mut references` (e.g. the `LintFormatter` having two params because one of them is invariant).

Reply:
Note that the way we use it already involves different lifetimes -- the content linters are within the scope of the `lint::run` function but the path isn't.

Reply:
Ah yes you're right, and thanks for the pointer :)